File: execution/directCsvImport.py
#!/usr/bin/env python3
"""
Streaming CSV → SQLite importer for large CA unclaimed property files.
Uses Python's csv.DictReader (fast C-backed) + sqlite3 batch inserts.
Handles the CA-specific column mapping.

Usage: python3 directCsvImport.py "/path/to/From_500_To_Beyond_1*.csv"
"""

import csv
import sqlite3
import sys
import os
import glob
import time

DB_PATH = os.path.join(os.path.dirname(os.path.abspath(__file__)), "..", "data.sqlite")
BATCH_SIZE = 100000

def import_csv(filepath):
    print(f"\n=== {os.path.basename(filepath)} ===")
    size_mb = os.path.getsize(filepath) / 1_048_576
    print(f"  Size: {size_mb:.1f} MB")

    db = sqlite3.connect(DB_PATH)
    db.execute("PRAGMA journal_mode = WAL")
    db.execute("PRAGMA synchronous = OFF")
    db.execute("PRAGMA cache_size = -65536")
    db.execute("PRAGMA foreign_keys = ON")

    # Deduplication query for fast count
    existing = db.execute("SELECT COUNT(*) FROM assets").fetchone()[0]
    source_url = f"upload:{os.path.basename(filepath)}"

    batch = []
    imported = 0
    skipped = 0
    row_count = 0
    start = time.time()

    with open(filepath, "r", encoding="utf-8", errors="replace") as f:
        reader = csv.DictReader(f)

        for row in reader:
            owner_name = row.get("OWNER_NAME", "").strip() or row.get("OWNER_NAME", "")
            owner_state = row.get("OWNER_STATE", "").strip()
            cash_reported = row.get("CASH_REPORTED", "").strip()
            current_balance = row.get("CURRENT_CASH_BALANCE", "").strip()
            property_type = row.get("PROPERTY_TYPE", "").strip() or None
            holder_name = row.get("HOLDER_NAME", "").strip() or None
            owner_city = row.get("OWNER_CITY", "").strip()
            owner_street = row.get("OWNER_STREET_1", "").strip()
            owner_zip = row.get("OWNER_ZIP", "").strip()
            property_id = row.get("PROPERTY_ID", "").strip() or None
            shares = row.get("SHARES_REPORTED", "").strip()
            securities_name = row.get("NAME_OF_SECURITIES_REPORTED", "").strip()

            if not owner_name:
                owner_name = "Unknown"

            if not owner_name or owner_name == "Unknown" and not owner_state:
                skipped += 1
                continue

            # Parse amount: prefer CASH_REPORTED, fall back to CURRENT_CASH_BALANCE
            amount = 0
            for val in [cash_reported, current_balance]:
                if val:
                    try:
                        cleaned = val.replace("$", "").replace(",", "")
                        amount = float(cleaned)
                        break
                    except ValueError:
                        pass

            # Build location
            parts = []
            if owner_street:
                parts.append(owner_street)
            if owner_city:
                parts.append(owner_city)
            if owner_state:
                parts.append(owner_state)
            if owner_zip:
                parts.append(owner_zip)
            location = ", ".join(parts) if parts else None

            # For property_type, append securities info if available
            if securities_name and property_type:
                property_type = f"{property_type} | {securities_name}"
            elif securities_name:
                property_type = securities_name

            batch.append((
                owner_name, None, None, owner_state,
                property_type, amount, holder_name,
                location, property_id, source_url, "manual_entry"
            ))

            row_count += 1

            if len(batch) >= BATCH_SIZE:
                db.executemany(
                    """INSERT OR IGNORE INTO assets
                       (owner_name, first_name, last_name, state, property_type, amount,
                        company, location, state_id, source_url, confidence)
                       VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)""",
                    batch
                )
                imported += len(batch)
                elapsed = time.time() - start
                rate = row_count / elapsed if elapsed > 0 else 0
                print(f"  [{elapsed:.0f}s] {row_count:,} rows | {imported:,} imported | {rate:,.0f} rows/sec")
                batch = []

    # Flush remaining
    if batch:
        db.executemany(
            """INSERT OR IGNORE INTO assets
               (owner_name, first_name, last_name, state, property_type, amount,
                company, location, state_id, source_url, confidence)
               VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)""",
            batch
        )
        imported += len(batch)

    total = time.time() - start
    total_imported = db.execute("SELECT COUNT(*) FROM assets").fetchone()[0]
    print(f"  DONE: {row_count:,} rows parsed, {imported:,} newly imported")
    print(f"  Total DB: {total_imported:,} assets | Time: {total:.1f}s | Rate: {row_count/total:,.0f} rows/sec")
    db.close()


if __name__ == "__main__":
    args = sys.argv[1:]
    if not args:
        print("Usage: python3 directCsvImport.py <csv_file_or_glob...>")
        sys.exit(1)

    files = []
    for arg in args:
        matched = glob.glob(arg)
        if matched:
            files.extend(matched)
        elif os.path.isfile(arg):
            files.append(arg)
        else:
            print(f"Not found: {arg}")

    print(f"Files to process: {len(files)}")
    for f in files:
        import_csv(f)

    print("\n=== All imports complete ===")

File: execution/directCsvImport.ts
/**
 * STREAMING CSV → SQLite importer for large CA unclaimed property files.
 * Uses csv-parse streaming to avoid OOM on 267MB files.
 *
 * Usage: npx tsx execution/directCsvImport.ts "/path/to/From_500_To_Beyond_*.csv"
 */

import Database from "better-sqlite3";
import { parse } from "csv-parse";
import * as fs from "fs";
import * as path from "path";
import { fileURLToPath } from "url";
import { Transform, TransformCallback } from "stream";
import {
  buildLocation,
  detectColumnMapping,
  parseAmount,
  pickFirst,
  type CsvColumnMapping,
} from "./csvAssetMapping";

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);
const DB_PATH = path.resolve(__dirname, "..", "data.sqlite");
const BATCH_SIZE = 5000;

function importCsv(filePath: string): Promise<void> {
  return new Promise((resolve, reject) => {
    console.log(`\n=== Reading: ${filePath} ===`);
    const stat = fs.statSync(filePath);
    console.log(`  Size: ${(stat.size / 1_048_576).toFixed(1)} MB`);

    const filename = path.basename(filePath);
    const sourceUrl = `upload:${filename}`;

    const db = new Database(DB_PATH);
    db.pragma("foreign_keys = ON");
    db.pragma("journal_mode = WAL");
    db.pragma("synchronous = OFF");
    db.pragma("cache_size = -65536");

    const stmt = db.prepare(`
      INSERT OR IGNORE INTO assets (owner_name, first_name, last_name, state, property_type, amount, company, location, state_id, source_url, confidence)
      VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
    `);

    let imported = 0;
    let skipped = 0;
    let rowCount = 0;
    let colMap: CsvColumnMapping | null = null;
    let batch: Record<string, string>[] = [];
    const startTime = Date.now();

    const insertBatch = db.transaction((rows: Record<string, string>[]) => {
      for (const row of rows) {
        const ownerName = pickFirst(row, colMap!.ownerName) || "Unknown";
        const state = pickFirst(row, colMap!.state) || "";
        const propertyType = pickFirst(row, colMap!.propertyType) || undefined;
        const amount = parseAmount(pickFirst(row, colMap!.amount));
        const company = pickFirst(row, colMap!.company) || undefined;
        const location = buildLocation(row, colMap!);
        const stateId = pickFirst(row, colMap!.stateId) || undefined;

        if (!state && ownerName === "Unknown") {
          skipped++;
          continue;
        }

        stmt.run(
          ownerName,
          undefined,
          undefined,
          state,
          propertyType,
          amount,
          company,
          location,
          stateId,
          sourceUrl,
          "manual_entry"
        );
        imported++;
      }
    });

    const readStream = fs.createReadStream(filePath, { encoding: "utf-8" });

    const parser = parse({
      columns: true,
      skip_empty_lines: true,
      trim: true,
      relax_column_count: true,
      relax_quotes: true,
    });

    // Buffer rows into batches
    const batcher = new Transform({
      objectMode: true,
      transform(row: Record<string, string>, _encoding: BufferEncoding, callback: TransformCallback) {
        if (!colMap) {
          const headers = Object.keys(row);
          colMap = detectColumnMapping(headers);
          console.log(`  Columns: ${headers.length}, mapping:`);
          console.log(`    ownerName: ${colMap.ownerName}`);
          console.log(`    state: ${colMap.state}`);
          console.log(`    amount: ${colMap.amount}`);
          console.log(`    company: ${colMap.company}`);
        }

        batch.push(row);
        rowCount++;

        if (batch.length >= BATCH_SIZE) {
          insertBatch(batch);
          batch = [];
          const pct = (rowCount / 1000000 * 100).toFixed(1); // rough % based on 1M rows per file
          const elapsed = ((Date.now() - startTime) / 1000).toFixed(0);
          process.stdout.write(`\r  [${elapsed}s] ${rowCount.toLocaleString()} rows, ${imported.toLocaleString()} imported...`);
        }
        callback();
      },
    });

    parser.on("error", (err: Error) => {
      reject(err);
    });

    readStream.on("error", (err: Error) => {
      reject(err);
    });

    readStream
      .pipe(parser)
      .pipe(batcher)
      .on("finish", () => {
        // Flush remaining batch
        if (batch.length > 0) {
          insertBatch(batch);
        }

        const totalSec = ((Date.now() - startTime) / 1000).toFixed(1);
        console.log(`\r  Done: ${rowCount.toLocaleString()} rows, ${imported.toLocaleString()} imported, ${skipped} skipped in ${totalSec}s`);
        db.close();
        resolve();
      })
      .on("error", (err: Error) => {
        reject(err);
      });
  });
}

// Main
const args = process.argv.slice(2);

if (args.length === 0) {
  console.error("Usage: npx tsx execution/directCsvImport.ts <csv_file_or_glob...>");
  process.exit(1);
}

(async () => {
  for (const arg of args) {
    if (arg.includes("*")) {
      const { execSync } = await import("child_process");
      const result = execSync(`ls ${arg} 2>/dev/null`, { encoding: "utf-8" });
      const files = result.trim().split("\n").filter(Boolean);
      for (const file of files) await importCsv(file);
    } else if (fs.existsSync(arg)) {
      await importCsv(arg);
    } else {
      console.error(`File not found: ${arg}`);
    }
  }
  console.log("\n=== All imports complete ===");
})();
